tick/optim/solver/src/svrg.cpp

```diff
@@ -1,127 +1,257 @@
-            //
-            iterate_no_interc.mult_incr(x_i, delta);
-            iterate[n_features] += delta;
-            iterate.mult_incr(mu, -step);
-        } else {
-            iterate.mult_incr(x_i, delta);
-            iterate.mult_incr(mu, -step);
+void SVRG::solve_sparse_exact_updates() {
+  // Data is sparse, and we use the exact strategy based on delayed updates.
+  // This means that the model is a child of ModelGeneralizedLinear.
+  // The strategy used here uses delayed updates strategy and the
+  // delayed penalization trick when the prox is separable, so that
+  // we can work when possible only inside the current support
+  // (non-zero values) of the sampled vector of features
+  for (t = 0; t < epoch_size; ++t) {
+    // Get next sample index
+    ulong i = get_next_i();
+    // Sparse features vector
+    BaseArrayDouble x_i = model->get_features(i);
+    // Gradients factors (model is a GLM)
+    // TODO: a grad_i_factor(i, array1, array2) to loop once on the features
+    double grad_i_diff = model->grad_i_factor(i, iterate) - model->grad_i_factor(i, fixed_w);
+    // We update the iterate within the support of the features vector
+    for (ulong idx_nnz = 0; idx_nnz < x_i.size_sparse(); ++idx_nnz) {
+      // Get the index of the idx-th sparse feature of x_i
+      ulong j = x_i.indices()[idx_nnz];
+      // How many iterations since the last update of feature j ?
+      ulong delay_j = 0;
+      ulong last_time_j_plus_one = last_time[j] + 1;
+      if (t > last_time_j_plus_one) {
+        delay_j = t - last_time_j_plus_one;
+      }
+      // Full gradient's coordinate j
+      double full_gradient_j = full_gradient[j];
+      // If there is delay
+      if (delay_j > 0) {
+        // then we need to correct variance reduction
+        iterate[j] -= step * delay_j * full_gradient_j;
+        // and regularization if prox is separable
+        if (is_prox_separable) {
+          iterate[j] = casted_prox->call_single(iterate[j], step, delay_j);
         }
-
-        prox->call(iterate, step, iterate);
-
-        if (variance_reduction == VarianceReductionMethod::Random && t == rand_index)
-            next_iterate = iterate;
-
-        if (variance_reduction == VarianceReductionMethod::Average)
-            next_iterate.mult_incr(iterate, 1.0 / epoch_size);
+      }
+      // Apply gradient descent to the model weights in the support of x_i
+      iterate[j] -= step * (x_i.data()[idx_nnz] * grad_i_diff + full_gradient_j);
+      if (is_prox_separable) {
+        // And regularize if the prox is separable
+        iterate[j] = casted_prox->call_single(iterate[j], step);
+      }
+      // Update last_time
+      last_time[j] = t;
     }
-
-    if (variance_reduction == VarianceReductionMethod::Last)
-        next_iterate = iterate;
+    if (!is_prox_separable) {
+      // If prox is not separable, apply it in a non-delayed fashion
+      prox->call(iterate, step, iterate);
```

Comment:
Is this correct?

I think this will apply a prox on all coeffs of iterate but only coeffs that are in the support of x_i are would be up to date.